crypto/hash/hmac_ossl.c

```diff
@@ -118,38 +184,66 @@ static srtp_err_status_t srtp_hmac_dealloc(srtp_auth_t *a)
 
 static srtp_err_status_t srtp_hmac_start(void *statev)
 {
-    HMAC_CTX *state = (HMAC_CTX *)statev;
-
-    if (HMAC_Init_ex(state, NULL, 0, NULL, NULL) == 0)
+    srtp_hmac_ossl_ctx_t *hmac = (srtp_hmac_ossl_ctx_t *)statev;
+
+#ifdef SRTP_OSSL_USE_EVP_MAC
+    if (hmac->use_dup) {
+        if (hmac->ctx_dup) {
+            EVP_MAC_CTX_free(hmac->ctx_dup);
+        }
+        hmac->ctx_dup = EVP_MAC_CTX_dup(hmac->ctx);
+        if (hmac->ctx_dup == NULL) {
+            return srtp_err_status_alloc_fail;
+        }
+    } else {
+        if (EVP_MAC_init(hmac->ctx, NULL, 0, NULL) == 0)
+            return srtp_err_status_auth_fail;
+    }
+#else
+    if (HMAC_Init_ex(hmac->ctx, NULL, 0, NULL, NULL) == 0)
         return srtp_err_status_auth_fail;
-
+#endif
     return srtp_err_status_ok;
 }
 
 static srtp_err_status_t srtp_hmac_init(void *statev,
                                         const uint8_t *key,
                                         int key_len)
 {
-    HMAC_CTX *state = (HMAC_CTX *)statev;
+    srtp_hmac_ossl_ctx_t *hmac = (srtp_hmac_ossl_ctx_t *)statev;
 
-    if (HMAC_Init_ex(state, key, key_len, EVP_sha1(), NULL) == 0)
-        return srtp_err_status_auth_fail;
+#ifdef SRTP_OSSL_USE_EVP_MAC
+    OSSL_PARAM params[2];
 
+    params[0] = OSSL_PARAM_construct_utf8_string("digest", "SHA1", 0);
+    params[1] = OSSL_PARAM_construct_end();
+
+    if (EVP_MAC_init(hmac->ctx, key, key_len, params) == 0)
+        return srtp_err_status_auth_fail;
+#else
+    if (HMAC_Init_ex(hmac->ctx, key, key_len, EVP_sha1(), NULL) == 0)
+        return srtp_err_status_auth_fail;
+#endif
     return srtp_err_status_ok;
 }
 
 static srtp_err_status_t srtp_hmac_update(void *statev,
                                           const uint8_t *message,
                                           int msg_octets)
 {
-    HMAC_CTX *state = (HMAC_CTX *)statev;
+    srtp_hmac_ossl_ctx_t *hmac = (srtp_hmac_ossl_ctx_t *)statev;
 
     debug_print(srtp_mod_hmac, "input: %s",
                 srtp_octet_string_hex_string(message, msg_octets));
 
-    if (HMAC_Update(state, message, msg_octets) == 0)
+#ifdef SRTP_OSSL_USE_EVP_MAC
+    if (EVP_MAC_update(hmac->use_dup ? hmac->ctx_dup : hmac->ctx, message,
+                       msg_octets) == 0)
```

Comment:
```suggestion
    if (EVP_MAC_update(hmac->ctx, message, msg_octets) == 0)
```

(Goes with the suggestion above to swap the roles of `ctx` and `ctx_dup`.)